src/main/java/io/cdap/plugin/servicenow/sink/service/ServiceNowSinkAPIRequestImpl.java

```diff
@@ -0,0 +1,259 @@
+            .getAsInt() == HttpStatus.SC_FORBIDDEN) {
+            String encodedResponseBody = servicedRequestsArray.get(i).getAsJsonObject().get(ServiceNowConstants.BODY)
+              .getAsString();
+            String decodedResponseBody = new String(Base64.getDecoder().decode(encodedResponseBody));
+            String errorDetail = jsonParser.parse(decodedResponseBody).getAsJsonObject().get(ServiceNowConstants.ERROR)
+              .getAsJsonObject().get(ServiceNowConstants.ERROR_DETAIL).getAsString();
+            if (errorDetail.equals(ServiceNowConstants.ACL_EXCEPTION)) {
+              throw new IllegalStateException(String.format("Permission denied for '%s' operation.",
+                config.getOperation()));
+            } else {
+              LOG.warn("Error Response : {} ", decodedResponseBody);
+            }
+          }
+        }
+
+        JsonArray unservicedRequestsArray = responseJSON.get(ServiceNowConstants.UNSERVICED_REQUESTS).getAsJsonArray();
+        if (unservicedRequestsArray.size() > 0) {
+          // Process unserviced requests array into unserviced requests list
+          List<RestRequest> unservicedRequests = processUnservicedRequestsArray(restRequestList,
+                                                                                unservicedRequestsArray);
+          // Retry unserviced requests
+          createPostRequest(restRequestList, unservicedRequests, accessToken);
+        }
+      }
+    } catch (IOException e) {
+      LOG.error("Error while connecting to ServiceNow", e.getMessage());
+      throw new RetryableException();
+    } catch (Exception e) {
+      throw e;
+    }
+  }
+
+  private String getErrorMessage(String responseBody) {
+    try {
+      JsonObject jo = gson.fromJson(responseBody, JsonObject.class);
+      return jo.getAsJsonObject(ServiceNowConstants.ERROR).get(ServiceNowConstants.MESSAGE).getAsString();
+    } catch (Exception e) {
+      return e.getMessage();
+    }
+  }
+
+  private ServiceNowBatchRequest getPayloadRequest(List<RestRequest> restRequests) {
+    ServiceNowBatchRequest payloadRequest = new ServiceNowBatchRequest();
+    payloadRequest.setBatchRequestId(UUID.randomUUID().toString());
+    payloadRequest.setRestRequests(restRequests);
+    return payloadRequest;
+  }
+
+  /**
+   * Process the unserviced requests array into unserviced requests.    .
+   *
+   * @param restRequestList         The list of rest Requests
+   * @param unservicedRequestsArray An array of unserviced requests
+   */
+  private List<RestRequest> processUnservicedRequestsArray(List<RestRequest> restRequestList,
+                                                           JsonArray unservicedRequestsArray)
+    throws InterruptedException, OAuthProblemException {
+    List<RestRequest> unservicedRequests = new ArrayList<>();
+    List<Integer> unservicedRequestsIds = new ArrayList();
+    for (int i = 0; i < unservicedRequestsArray.size(); i++) {
```

Comment:
create a Map request id-> RestRequest and lookup ids in unservicedRequestsArray in the map.

Reply:
refactored the logic by replacing the list with a map. bc67f6e1b45ebd37026d24f7b64a2d8d43fbbc88